proposals/0005-fallback-rounting.md

```diff
@@ -0,0 +1,229 @@
+- **Priority-based Routing** e.g. OpenAI should serve all requests if available. If not, then Anthropic, and finally, Google Bard
+- **Latency-based Routing** e.g. the next request should be served by the fastest available model in the pool. This is useful for real-time applications where latency is important.
+- **Cost-based Routing** e.g. the next request should be served by the cheapest available model in the pool
+- **Performance-based Routing** e.g. the next request should be served by the model that performs the best for my request (see GEP003).
+- etc.
+
+WRR could be useful to perform **A/B testing** of model pool setup to find the best model/params/prompt combination for a given task.
+
+### Requirements
+
+- R1: We MUST fallback to a healthy model if any in a pool doing our best to fulfill routing strategy specified in configs
+- R2: Fallback MUST be as quick as possible. No redundant wasteful retrying or waiting
+
+## Design
+
+The off-the-shelf strategy to tackle the availability issues is redundancy.
+So we have some backup models to talk to in case the main one is down.
+The mechanism of switching models in case of failures is called fallbacking.
+
+For most serious production applications that have real users, downtime is not an option. Hence, fallbacking should be
+an intrinsic feature of Glide routing despite routing strategy.
+
+There may be exceptions to this dictated by the model nature.
+For example, embedding API may require one model to be used for all embeddings produced.
+
+Based on this, we group all models we support into types (see GEP0002):
+
+- language models
+- embeddings
+- speech transcribers
+- speech synthesizers
+- etc.
+
+Each model type is represented as a list of model pools that could serve the incoming request uniformly.
+Each model pool is a list of models (along with their params & operational info like API key) and routing strategy that defines
+how incoming requests should be served.
+
+The model pool ID or name is exposed to clients and they provide IDs during requests (see GEP0002).
+This is how Glide understands which one should serve the concrete request.
+
+### Model Routing & Fallbacking
+
+Glide leverages fallback models on the first fail and do adaptive health tracking (via token bucket). 
+Moving right to the next model serves as a retry for the first one while we spend as little time as possible waiting. 
+
+Then, the whole service instance is going to be aware about number of fails that happen to a certain model 
+and as soon as that number goes over the roof, 
+the model is going to be "disqualified" to serve requests for some time. 
+This means that as soon as the model is dead we pull it out of the pool, so the next requests are going to be served by healthy models only. 
+This strategy should lead to much better latency on failover.
+
+![Routing & Fallbacks](./imgs/glide-fallback-logic.png)
+
+### Routing Strategies
+
+#### Priority
```

Comment:
Does each router have its own token bucket? Or is there a single shared bucket across the entire deployment?

Reply:
@mkrueger12 In the current implementation, buckets are specific for routers. Ideally, we need to share the health of providers across the whole Glide cluster not only deployment. So yeah, you are asking a great question here.

What would be a good identifier for the provider/model given on-prem providers? Provider's Base URL?

Reply:
Filled a ticket not to forget about this: https://github.com/EinStack/glide/issues/87

Reply:
Yeah probably the providers base url but I'm not entirely sure to be honest.